File: term1/hw/program_design/100_chickens.py
lst = [0 for i in range(1_000)]
def purchase(money, number, types, prices):
    if  money < 0 or (not prices) or types <= 0:
        return
    if  money == 0 and number == 0:
        for i in range(len(prices)):
            print(lst[i], end=" " if i != len(prices) - 1 else '\n')
        return

    purchase(money-prices[len(prices) - types], number-1, types, prices)
    lst[len(prices) - types] += 1
    purchase(money, number, types-1, prices)

def main():
    prices = [15, 9, 1]
    purchase(300, 100, 3, prices)
    return 0

if __name__ == '__main__':
    main()

File: term1/hw/program_design/100_chickens.cpp
#include <iostream>

void purchaseHelper(int money, int number, int types, const int prices[], int currentTypes, int currentCounts[], int index) {
    if (index == types) {
        // Base case: reached the end of the prices array
        if (number == 0 && money == 0) {
            // Output the solution
            for (int i = 0; i < types; ++i) {
                std::cout << currentCounts[i] << " ";
            }
            std::cout << std::endl;
        }
        return;
    }

    // Try buying different amounts of current type of chicken
    for (int count = 0; count <= number; ++count) {
        int cost = count * prices[index];
        if (cost <= money) {
            // Update current counts and call recursively for the next type
            currentCounts[currentTypes] = count;
            purchaseHelper(money - cost, number - count, types, prices, currentTypes + 1, currentCounts, index + 1);
        }
    }
}

void purchase(int money, int number, int types, const int prices[]) {
    // Initialize an array to store the counts of each type of chicken
    int currentCounts[types] = {0};
    // Start the recursive process
    purchaseHelper(money, number, types, prices, 0, currentCounts, 0);
}

int main() {
    const int prices[] = {15, 9, 1};
    purchase(300, 100, 3, prices);
    return 0;
}

